peers/app_request_network.go

```diff
@@ -126,95 +133,155 @@ func NewNetwork(
+			break
+		}
+	}
+
 	go logger.RecoverAndPanic(func() {
 		testNetwork.Dispatch()
 	})
 
+	arNetwork := &appRequestNetwork{
+		network:         testNetwork,
+		handler:         handler,
+		infoAPI:         infoAPI,
+		logger:          logger,
+		lock:            new(sync.RWMutex),
+		validatorClient: validatorClient,
+		metrics:         metrics,
+		trackedSubnets:  trackedSubnets,
+		manager:         manager,
+	}
+
+	arNetwork.startUpdateValidators()
+
 	return arNetwork, nil
 }
 
-// ConnectPeers connects the network to peers with the given nodeIDs.
-// Returns the set of nodeIDs that were successfully connected to.
-func (n *appRequestNetwork) ConnectPeers(nodeIDs set.Set[ids.NodeID]) set.Set[ids.NodeID] {
+func (n *appRequestNetwork) TrackSubnet(subnetID ids.ID) {
+	n.lock.RLock()
+	if n.trackedSubnets.Contains(subnetID) {
+		n.lock.RUnlock()
+		return
+	}
+	n.lock.RUnlock()
+
 	n.lock.Lock()
 	defer n.lock.Unlock()
+	n.logger.Debug("Tracking subnet", zap.String("subnetID", subnetID.String()))
+	n.trackedSubnets.Add(subnetID)
+	updateValidatorSet(context.Background(), n.logger, subnetID, n.manager, n.validatorClient)
+}
 
-	// First, check if we are already connected to all the peers
-	connectedPeers := n.network.PeerInfo(nodeIDs.List())
-	if len(connectedPeers) == nodeIDs.Len() {
-		return nodeIDs
-	}
+func (n *appRequestNetwork) startUpdateValidators() {
+	go func() {
+		// Fetch validators immediately when called, and refresh every ValidatorRefreshPeriod
+		ticker := time.NewTicker(ValidatorRefreshPeriod)
+		for ; true; <-ticker.C {
+			n.lock.Lock()
 
-	// If we are not connected to all the peers already, then we have to iterate
-	// through the full list of peers obtained from the info API. Rather than iterating
-	// through connectedPeers for already tracked peers, just iterate through the full list,
-	// re-adding connections to already tracked peers.
+			updateValidatorSet(context.Background(), n.logger, constants.PrimaryNetworkID, n.manager, n.validatorClient)
+			for _, subnet := range n.trackedSubnets.List() {
+				updateValidatorSet(context.Background(), n.logger, subnet, n.manager, n.validatorClient)
+			}
+			n.lock.Unlock()
+		}
+	}()
+}
 
-	startInfoAPICall := time.Now()
-	// Get the list of publicly discoverable peers
-	peers, err := n.infoAPI.Peers(context.Background(), nil)
-	n.setInfoAPICallLatencyMS(float64(time.Since(startInfoAPICall).Milliseconds()))
+func updateValidatorSet(
+	ctx context.Context,
+	logger logging.Logger,
+	subnetID ids.ID,
+	manager vdrs.Manager,
+	client *validators.CanonicalValidatorClient,
+) error {
+	logger.Debug("Fetching validators for subnet ID", zap.Stringer("subnetID", subnetID))
+
+	// Fetch the primary network validators from the P-Chain
+	height, err := client.GetCurrentHeight(ctx)
 	if err != nil {
-		n.logger.Error(
-			"Failed to get peers",
-			zap.Error(err),
-		)
-		return nil
+		return err
+	}
+	validators, err := client.GetValidatorSet(ctx, height, subnetID)
```

Comment:
this won't work publicly with a non-"proposed" height though right?

Reply:
Good point. I updated this to call `getValidatorsAt(proposed)` under the hood.